toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/examples/test_xnvme_hello.py

```diff
@@ -0,0 +1,10 @@
+import pytest
+from cijoe.xnvme.tests.conftest import xnvme_device_driver as device
+from cijoe.xnvme.tests.conftest import xnvme_parametrize
```

Comment:
I recall having issues with that, which i think is related to:
https://github.com/pytest-dev/pytest/issues/1596

However, since the intent is enabling easy deployment of the tests and running them, via "pytest --pyargs cijoe.xnvme.tests". Not so much the inplace-edit-and-run. Then i didnt pursue it any further.

I don't see saving a single line of import atatement per test-module as a reason to not go ahead with merging. Wouldn't you agree?

Reply:
To elaborate a bit on this, for in-place-edit-and run, one can set the ``PYTHONPATH``, for example:

	PYTHONPATH="$HOME/git/xnvme/toolbox/cijoe-pkg-xnvme/src" pytest --pyargs cijoe.xnvme.tests -k info ...

Then Python will find the non-installed package before any installed packages. Doing so enables the in-place-edit-and-run. The above can also be used, in case one just does not want to install the package.

Another thing that I have found myself doing is, writing a test, say ``test_foo.py``:

	from cijoe.xnvme.tests.conftest import xnvme_device_driver as device
	from cijoe.xnvme.tests.conftest import xnvme_parametrize

	@xnvme_parametrize(labels=[], opts=["be", "sync", "async"])
	def test_foo(cijoe, device, be_opts, cli_args):
	    assert True

And having that test in an arbitrary "out-of-tree" location, that is, not a child of "cijoe-pkg-xnvme/src/cijoe/xnvme/tests". And running it:

	pytest test_foo.py ...

Since it is "out-of-tree" the fully qualified imports are needed. Once I am done with the test I copy it into the python-package without any need to change the imports. Similarly, one can copy any of the tests "in-tree" to an arbitrary "out-of-tree" location and run it "standalone" without changing the imports.

So, to address whether the fully-qualified import paths should be made relative, I would prefer not to, because:

* They can be run 'in-place' by setting the PYTHONPATH
* Tests can be moved in and out of "tree" without having to change the imports

I find the latter really convenient.

Regarding the import of "xnvme_device_driver as device", i agree that it would be nice if it could be avoided. However, it seems like a deficiency of pytest-collection, not really something we can affect without making other sacrifices.

Reply:
> the intent is enabling easy deployment of the tests and running them, via "pytest --pyargs cijoe.xnvme.tests". Not so much the inplace-edit-and-run.

In what cases would we install the package instead of running it from the source? When developing, and deploying to GHA, it would seem easiest not to have to build the package and install it first. And what I'm proposing won't block you from doing it anyway.

> I don't see saving a single line of import atatement per test-module as a reason to not go ahead with merging. Wouldn't you agree?

It's not about have an import statement more or less, it's the underlaying issue that concerns me. We need to find the root cause and either say we accept it, or fix it.


> one can set the PYTHONPATH

That is a possible workaround, but it also seems like a lot of added complexity, for something that should just work.

> having that test in an arbitrary "out-of-tree" location, that is, not a child of "cijoe-pkg-xnvme/src/cijoe/xnvme/tests".

Is that a plausible scenario? Wouldn't you be developing the tests and store them all in the same place? Or what would the benefit of that be?

> However, it seems like a deficiency of pytest-collection

That again seems to be a symptom of the root cause I want to avoid. The `conftest.py` is not properly located during collection. It should be an easy thing to fix.

Reply:
> > the intent is enabling easy deployment of the tests and running them, via "pytest --pyargs cijoe.xnvme.tests". Not so much the inplace-edit-and-run.
> 
> In what cases would we install the package instead of running it from the source? When developing, and deploying to GHA, it would seem easiest not to have to build the package and install it first. And what I'm proposing won't block you from doing it anyway.
> 

This is how it is being used now. It is fairly convenient to be able to do ``pytest --pyargs cijoe.xnvme.tests``  instead of relying on the availability of the source repository and knowing where it is located. This usage takes care of it. 

> > I don't see saving a single line of import atatement per test-module as a reason to not go ahead with merging. Wouldn't you agree?
> 
> It's not about have an import statement more or less, it's the underlaying issue that concerns me. We need to find the root cause and either say we accept it, or fix it.
> 

As mentioned earlier, it seems to me that it is a long-lived pytest-deficiency https://github.com/pytest-dev/pytest/issues/1596
Going back to 2016. I am fine with it, the functionality introduced with this PR works as intended and it is a vast improvement on the status quo.

> > one can set the PYTHONPATH
> 
> That is a possible workaround, but it also seems like a lot of added complexity, for something that should just work.
> 
> > having that test in an arbitrary "out-of-tree" location, that is, not a child of "cijoe-pkg-xnvme/src/cijoe/xnvme/tests".
> 
> Is that a plausible scenario? Wouldn't you be developing the tests and store them all in the same place? Or what would the benefit of that be?
> 

I have been doing it, that is why I mention it.

> > However, it seems like a deficiency of pytest-collection
> 
> That again seems to be a symptom of the root cause I want to avoid. The `conftest.py` is not properly located during collection. It should be an easy thing to fix.

If it is an easy thing to fix, then please send a PR after this fixing it. Regardless, the only impact it will have on the functionality of this PR is whether or not those import statements are needed.

Reply:
Here is another perspective, before this PR there was **no** way of running the tests without installing the package.
At least now, there **is** a way, by setting ``PYTHONPATH`` and running them directly from the repository.

Reply:
@Baekalfen thanks for the offline discussion it helped me understand your perspective better.

If you have the time, then have a look a the ``revised: ...`` commit. Briefly:

* Imports are now relative instead of absolute.

* Inside the conf-test the ``xnvme_device_driver()`` is now renamed to
``device()`` to satisfy the pytest.parametrize/indirect. Thus the
``xnvme_device_driver as device`` import is no longer needed and removed.

* One can now run the tests in-place **without** having to specify ``PYTHONPATH`` and most importantly **without** having to install the package.

* Running it as a package still works.

Reply:
Also, just added a brief description on how to run pytest directly, that is, without the use of **cijoe**.

Reply:
Note, "revised" are gone as the history has been cleaned up.